railroad.py

```diff
@@ -1426,18 +1864,433 @@ def format(self, x: float, y: float, width: float) -> Skip:
         Path(x, y).right(width).addTo(self)
         return self
 
+    def textDiagram(self) -> TextDiagram:
+        (line,) = TextDiagram._getParts(["line"])
+        return TextDiagram(0, 0, [line])
+
     def __repr__(self) -> str:
         return "Skip()"
 
 
+class TextDiagram:
+    # Characters to use in drawing diagrams.  See setFormatting(), PARTS_ASCII, and PARTS_UNICODE.
+    parts: Dict[str, str]
+
+    def __init__(self, entry: int, exit: int, lines: List[str]) -> TextDiagram:
+        # entry: The entry line for this diagram-part.
+        self.entry: int = entry
+        # exit: The exit line for this diagram-part.
+        self.exit: int = exit
+        # height: The height of this diagram-part, in lines.
+        self.height: int = len(lines)
+        # lines[]: The visual data of this diagram-part.  Each line must be the same length.
+        self.lines: List[str] = lines.copy()
+        # width: The width of this diagram-part, in character cells.
+        self.width: int = len(lines[0]) if len(lines) > 0 else 0
+        nl = "\n"  # f-strings can't contain \n until Python 3.12
```

Comment:
I don't know what you mean by this, `\n` is perfectly fine since f-strings were first introduced. The new 3.12 thing is that they actually use a real parser instead of hacking the existing string parser, so you can use the same quoting character inside of the the interpolation, like `f"foo {" ".join(bars)} baz"`. But `f"foo\nbar"` has always been valid.

Reply:
I misunderstood something I read.  The f-string change to a real grammar and parser affects more than just quotes (see PEP 701), and in particular, this misled me:

> Backslashes and unicode characters: before Python 3.12 f-string expressions couldn’t contain any \ character. This also affected unicode escape sequences (such as \N{snowman}) as these contain the \N part that previously could not be part of expression components of f-strings. 

(https://docs.python.org/3/whatsnew/3.12.html#whatsnew312-pep701)

But, of course, you're right.